### PR TITLE
[Doris On ES][Optimization] Optimize Doris On Elasticsearch performance using docvalues

### PR DESCRIPTION
**Pull request by wuyunfeng**
Pure DocValue optimization for doris-on-es

https://github.com/apache/incubator-doris/issues/2048

Future todo:

Today, for every tuple scan we check if pure_docvalue is enabled, this is not reasonable,  should check pure_docvalue enabled for one whole scan outside,  I will add this todo in future

### CODE REVIEW

File: fe/src/main/java/org/apache/doris/catalog/EsTable.java [hunk 1]
```diff
@@ -160,34 +225,60 @@ public int getSignature(int signatureVersion) {
...
+            }
         } else {
-            throw new IOException("invalid partition type: " + partType);
+            hosts = Text.readString(in);
+            seeds = hosts.split(",");
+            userName = Text.readString(in);
+            passwd = Text.readString(in);
+            indexName = Text.readString(in);
+            mappingType = Text.readString(in);
+            PartitionType partType = PartitionType.valueOf(Text.readString(in));
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/EsTable.java]: Don't you need to put these variables back to the tableContext?

File: fe/src/main/java/org/apache/doris/catalog/EsTable.java [hunk 2]
```diff
@@ -160,34 +225,60 @@ public int getSignature(int signatureVersion) {
...
-        userName = Text.readString(in);
-        passwd = Text.readString(in);
-        indexName = Text.readString(in);
-        mappingType = Text.readString(in);
-        PartitionType partType = PartitionType.valueOf(Text.readString(in));
-        if (partType == PartitionType.UNPARTITIONED) {
-            partitionInfo = SinglePartitionInfo.read(in);
-        } else if (partType == PartitionType.RANGE) {
-            partitionInfo = RangePartitionInfo.read(in);
+        if (Catalog.getCurrentCatalogJournalVersion() >= FeMetaVersion.VERSION_66) {
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/catalog/EsTable.java]: You need to change to VERSION_68 now

File: fe/src/main/java/org/apache/doris/external/EsStateStore.java [hunk 3]
```diff
@@ -17,6 +17,16 @@
 
 package org.apache.doris.external;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Range;
+import okhttp3.Authenticator;
+import okhttp3.Call;
+import okhttp3.Credentials;
```
Comment (morningman) [inline fe/src/main/java/org/apache/doris/external/EsStateStore.java]: import order should be:

org.apache.doris
com
org
java
javax

change the setting of your java IDE

**Review by morningman — CHANGES_REQUESTED** [review]

File: be/src/exec/es/es_scroll_parser.cpp [hunk 4]
```diff
@@ -233,17 +262,18 @@ int ScrollParser::get_total() {
...
     *line_eof = true;
     if (_size <= 0 || _line_index >= _size) {
         return Status::OK();
     }
 
     const rapidjson::Value& obj = _inner_hits_node[_line_index++];
-    const rapidjson::Value& line = obj[FIELD_SOURCE];
-    if (!line.IsObject()) {
-        return Status::InternalError("Parse inner hits failed");
+    bool pure_doc_value = false;
```
Comment (imay) [inline be/src/exec/es/es_scroll_parser.cpp]: we need check this every tuple? I think it should be checked once.

File: fe/src/main/java/org/apache/doris/catalog/EsTable.java [hunk 5]
```diff
@@ -160,34 +224,75 @@ public int getSignature(int signatureVersion) {
...
+
+            enableDocValueScan = Boolean.parseBoolean(tableContext.get("enableDocValueScan"));
+
+            PartitionType partType = PartitionType.valueOf(Text.readString(in));
+            if (partType == PartitionType.UNPARTITIONED) {
+                partitionInfo = SinglePartitionInfo.read(in);
+            } else if (partType == PartitionType.RANGE) {
+                partitionInfo = RangePartitionInfo.read(in);
+            } else {
+                throw new IOException("invalid partition type: " + partType);
```
[the remark this answers is not in the document]
Comment (morningman) [reply fe/src/main/java/org/apache/doris/catalog/EsTable.java]: This is ok, shouldn't let FE start if this happens

File: be/src/exec/es/es_scroll_parser.cpp (hunk 4 above)
Comment (wuyunfeng) [reply be/src/exec/es/es_scroll_parser.cpp]: Maybe need two function part if checked once

File: be/src/exec/es/es_scroll_query.cpp [hunk 6]
```diff
@@ -75,15 +74,32 @@ std::string ESScrollQueryBuilder::build(const std::map<std::string, std::string>
...
-    // just filter the selected fields for reducing the network cost
+    bool pure_docvalue = false;
+    // check docvalue sacan optimization
+    if (docvalue_context.size() > 0 && docvalue_context.size() >= fields.size()) {
+        for (auto iter = fields.begin(); iter != fields.end(); iter++) {
+            if (!(docvalue_context.find(*iter) != docvalue_context.end())) {
+                pure_docvalue = false;
+                break;
+            }
+            pure_docvalue = true;
```
[the remark this answers is not in the document]
Comment (wuyunfeng) [reply be/src/exec/es/es_scroll_query.cpp]: OK, the docvalue_context.size() > 0 would be check first.

File: be/src/exec/es/es_scroll_parser.cpp (hunk 4 above)
Comment (wuyunfeng) [reply be/src/exec/es/es_scroll_parser.cpp]: I would add this todo in future, currently modify this will consume mount of time because the code structure

**Review by morningman — DISMISSED** [review]
LGTM

File: be/src/exec/es/es_scroll_query.cpp [hunk 7]
```diff
@@ -75,15 +74,39 @@ std::string ESScrollQueryBuilder::build(const std::map<std::string, std::string>
...
-    // just filter the selected fields for reducing the network cost
-    if (fields.size() > 0) {
-        rapidjson::Value source_node(rapidjson::kArrayType);
+    bool pure_docvalue = true;
+    // check docvalue sacan optimization
+    if (docvalue_context.size() == 0 || docvalue_context.size() < fields.size()) {
+        pure_docvalue = false;
+    } else {
+        for (auto iter = fields.begin(); iter != fields.end(); iter++) {
+            if (!(docvalue_context.find(*iter) != docvalue_context.end())) {
```
Comment (imay) [inline be/src/exec/es/es_scroll_query.cpp]:
```suggestion
            if (docvalue_context.find(*iter) == docvalue_context.end()) {
```

File: be/src/exec/es/es_scroll_query.cpp [hunk 8]
```diff
@@ -75,15 +74,39 @@ std::string ESScrollQueryBuilder::build(const std::map<std::string, std::string>
...
     es_query_dsl.AddMember("query", query_node, allocator);
-    // just filter the selected fields for reducing the network cost
-    if (fields.size() > 0) {
-        rapidjson::Value source_node(rapidjson::kArrayType);
+    bool pure_docvalue = true;
+    // check docvalue sacan optimization
+    if (docvalue_context.size() == 0 || docvalue_context.size() < fields.size()) {
+        pure_docvalue = false;
+    } else {
+        for (auto iter = fields.begin(); iter != fields.end(); iter++) {
```
Comment (imay) [inline be/src/exec/es/es_scroll_query.cpp]:
```suggestion
        for (auto& field : fields) {
```

File: be/src/exec/es/es_scroll_query.cpp [hunk 9]
```diff
@@ -75,15 +74,39 @@ std::string ESScrollQueryBuilder::build(const std::map<std::string, std::string>
...
+            if (!(docvalue_context.find(*iter) != docvalue_context.end())) {
+                pure_docvalue = false;
+                break;
+            }
+        }
+    }
+    rapidjson::Value source_node(rapidjson::kArrayType);
+    if (pure_docvalue) {
+        for (auto iter = fields.begin(); iter != fields.end(); iter++) {
```
Comment (imay) [inline be/src/exec/es/es_scroll_query.cpp]:
```suggestion
        for (auto& field : fields) {
```

**Review by imay — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**